### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by RuiAAPeres** [thoughtbot/Argo#205]
I did a bit of digging in my app, to pinpoint from where the leak is coming from. I added a blob of `NSData` in my main bundle and did the following:

```
func testLeak() {
    let data = NSBundle.mainBundle().jsonData("blob")
    let parsedData = try! NSJSONSerialization.JSONObjectWithData(data!, options: NSJSONReadingOptions())
    let returnedJSON = JSON.parse(parsedData)
}
```

This is enough (even without using the variables) to create a leak. The Profile looks like the following:

<img width="1267" alt="screen shot 2015-09-02 at 19 39 11" src="https://cloud.githubusercontent.com/assets/949286/9640612/5f40d002-51aa-11e5-9be7-335f098b6318.png">

I did a mix of changes in the blob file, from Array, Strings, Bools, Ints, etc, to see if there was a specific case where the leak wouldn't come, but to no avail. It's always there (but ofc less things being leaked). For example doing this:

```
    let parsedData : AnyObject = "something"
    let returnedJSON = JSON.parse(parsedData)
```

Still leaks:

<img width="1270" alt="screen shot 2015-09-02 at 19 56 08" src="https://cloud.githubusercontent.com/assets/949286/9641081/b3b0940e-51ac-11e5-8df0-21844c3146b7.png">

Looking at your code, it doesn't seem to be any issue, so I searched a bit more and it seems there might be issues with Swift itself, like [this](https://forums.developer.apple.com/thread/4892) and [this](https://stackoverflow.com/questions/30628547/memory-leak-in-swift-structures-how-to-fix-this/31936407#31936407). In any case, I think this should be verified.

**Comment by gfontenot** [comment]
This is terrifying.

**Comment by RuiAAPeres** [comment]
@gfontenot a bug report has been filed to Apple.

**Comment by gfontenot** [comment]
@RuiAAPeres Thanks for taking care of that. Let me know if there's anything I can do on our end.

**Comment by jckarter** [comment]
This is our bug, sorry. The leak appears to be caused by a regression in `switch` codegen for `as` patterns. If I rewrite `JSON.parse` in terms of `if let`s, then I'm able to work around the leak, e.g.:

```swift
    static func parse(json: AnyObject) -> JSON {
      if let v = json as? Swift.String {
        return .String(v)
      }
      /* etc. */
    }
```

I'll let you know when we're able to get a fix out, but until then I think you'll need to adopt that as a workaround.

**Comment by jshier** [comment]
@RuiAAPeres Can you post your Apple bug number here? Future reports about the same issue (@jckarter, any plans to file?) should mention the original report.

**Comment by jckarter** [comment]
We have it as rdar://problem/22587077.

**Comment by RuiAAPeres** [comment]
@jckarter 👍

**Comment by paulyoung** [comment]
Is there a preferred branch to submit a PR against at this point?

**Comment by gfontenot** [comment]
@paulyoung I'd say `td-swift-2`.

**Comment by neilpa** [comment]
Looks like this isn't getting fixed in Xcode 7 since it's still listed as an issue in the 7.1 beta release notes

> Using switch against multiple types with as patterns may cause a memory leak. For example, avoid this kind of switch statement:
```swift
switch x {
 case let a as A: ...
 case let b as B: ...
 case let c as C: ...
 default: ...
}
```
Workaround: Rewrite the code to use `if let a = x as? A` statements instead of switch. This pattern performs type checks that avoid the leak. (22587077)

**Comment by gfontenot** [comment]
Yep, we'll watch the betas to see if this gets fixed. In the meantime, it might make sense to finish the work @paulyoung started in #217 and rewrite the rest of the switch statements. It'd probably be good to do the whole thing in one commit so that it can easily be reverted later.

Bummer.

**Comment by paulyoung** [comment]
@gfontenot I'll push the changes and squash the commits.

**Comment by tonyd256** [comment]
Fixed in #217
